src/debug_flags.cc

```diff
@@ -12,10 +12,35 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#include <stdio.h>
+#include <map>
+#include <vector>
+#include <string>
+
+#include "graph.h"
+
 bool g_explaining = false;
 
 bool g_keep_depfile = false;
 
 bool g_keep_rsp = false;
 
 bool g_experimental_statcache = true;
+
+// Reasons each Node needs rebuilding, for "-d explain".
+typedef std::map<const Node*, std::vector<std::string> > Explanations;
+static Explanations explanations_;
+
+void record_explanation(const Node* node, std::string explanation) {
```

Comment:
nit: use a `StringPiece` type instead of `std::string` here to avoid an un-necessary string copy, e.g.:

```
void record_explanation(const Node* node, StringPiece explanation) {
  explanations_[node].emplace_back(explanation.str_, explanation.len_);
}
```